### PR TITLE
fix Topological Sort

### PR DESCRIPTION
**Pull request by alimamdouh212**
#1337 fix

### CODE REVIEW

File: src/Traversals/depthfirst.jl [hunk 1]
```diff
@@ -95,14 +95,21 @@ end
...
     verts = Vector{T}()
     state = TopoSortState(vcolor, verts, zero(T))
-    for v in vertices(g)
-        state.vcolor[v] != 0 && continue
-        state.vcolor[v] = 1
-        if !traverse_graph!(g, v, DepthFirst(), state)
-            throw(CycleError())
-        end
+    
+    sources = filter(x -> isempty(inneighbors(g, x)), vertices(g))
```
Comment (sbromberger) [inline src/Traversals/depthfirst.jl]: So, the issue is that we can pass in a `neighborfn` to `DepthFirst()`. We need to use the inverse function here, so:
```
invdegreefn = alg.neighborfn == outneighbors ? indegree : outdegree
sources = filter(x -> iszero(invdegreefn(g, x)), vertices(g))
...
```

File: src/Traversals/depthfirst.jl [hunk 2]
```diff
@@ -95,14 +95,21 @@ end
...
+   
+   
+       
+    if !traverse_graph!(g, sources , DepthFirst(), state)
+         throw(CycleError())
+    end
+
```
Comment (sbromberger) [inline src/Traversals/depthfirst.jl]: We can shortcut this: `traverse_graph(g, sources, DepthFirst(), state) || throw(CycleError())`

File: src/Traversals/depthfirst.jl [hunk 3]
```diff
@@ -95,14 +95,21 @@ end
...
+    if !traverse_graph!(g, sources , DepthFirst(), state)
+         throw(CycleError())
+    end
+      
+    
+    if length(state.verts) == nv(g)
+        return reverse(state.verts)
+    else
+        throw(CycleError())
     end
```
Comment (sbromberger) [inline src/Traversals/depthfirst.jl]: Simplify this:
```
length(state.verts) == nv(g) || throw(CycleError())
return reverse!(state.verts)
```

**Comment by sbromberger** [comment]
@alimamdouh212 hold up a sec - I used your `filter` trick in #1343 :) Mind testing it?

Just add
```
indegree(g, v) != 0 && continue
```
right after
```
@inbounds for v in vertices(g)
```
in `toposort2()`.

**Comment by alimamdouh212** [comment]
@sbromberger  if the graph has multiple sources that will make the algorithm again discover visited nodes, the best solution is to make one dfs call

**Comment by sbromberger** [comment]
yeah, you're right. @abhinavmehndiratta came up with a good example. OK. Let's concentrate on #1340, and, failing that, #1341. I'll close this out.

**Comment by sbromberger** [comment]
Oops! I closed the wrong one. Sorry.

**Comment by sbromberger** [comment]
Can you confirm that you don't have the same issue with this graph?
```
Here you go -
julia> using LightGraphs, LightGraphs.Traversals

julia> g = SimpleDiGraph(5)
{5, 0} directed simple Int64 graph

julia> add_edge!(g, 1, 2)
true

julia> add_edge!(g, 2, 3)
true

julia> add_edge!(g, 2, 4)
true

julia> add_edge!(g, 3, 4)
true

julia> add_edge!(g, 5, 3)
true

julia> LightGraphs.Traversals.toposort2(g)
1 2
2 3
3 4
2 4
5 3
3 4   # right here. There's no reason to repeat the traversal here.
```

**Comment by alimamdouh212** [comment]
@sbromberger  sorry for being late, as it turns out, the current dfs can’t handle multiple sources, so i had to fix it, after that the example you gave worked fine, and i have put it in the tests.

File: src/Traversals/depthfirst.jl [hunk 4]
```diff
@@ -95,14 +95,13 @@ end
...
-            throw(CycleError())
-        end
-    end
-    return reverse(state.verts)
+    
+    sources = filter(x -> indegree(g, x) == 0, vertices(g))
+ 
+    traverse_graph!(g, sources, DepthFirst(), state) || throw(CycleError())
+      
+    length(state.verts) == nv(g) || throw(CycleError())
```
Comment (sbromberger) [inline src/Traversals/depthfirst.jl]: Let's be more specific here. If there's a cycle, will the length be greater than or less than `nv(g)`?

File: test/Traversals/depthfirst.jl [hunk 5]
```diff
@@ -3,7 +3,14 @@
     g5 = SimpleDiGraph(4)
     add_edge!(g5, 1, 2); add_edge!(g5, 2, 3); add_edge!(g5, 1, 3); add_edge!(g5, 3, 4)
     gx = cycle_digraph(3)
-  
+    gt = SimpleDiGraph(3)
```
Comment (sbromberger) [inline test/Traversals/depthfirst.jl]: Comment before this line explaining we're testing cycles.

Also - define these graphs immediately before the `@testset` where they're being used, not all up front like this.

File: src/Traversals/depthfirst.jl (hunk 4 above)
Comment (alimamdouh212) [reply src/Traversals/depthfirst.jl]: if there is a cycle that can’t be reached from any of the sources, it won’t be in state.verts, so if
 `length(state.verts) < nv(g)` that means there is a cycle

Comment (sbromberger) [reply src/Traversals/depthfirst.jl]: ok, let's make that explicit, and if the length is greater than throw a `TraversalError` of some sort.

Comment (alimamdouh212) [reply src/Traversals/depthfirst.jl]: won’t that make coverage issue if there is no test that will throws that error, can i make assert ?

Comment (sbromberger) [reply src/Traversals/depthfirst.jl]: That's the beauty of this: a short-circuiting statement gets evaluated so there's no coverage issue. That is,

```
true || throw(TraversalError())
```
will not fail code coverage. :)

**Comment by alimamdouh212** [comment]
@sbromberger  there is a coverage issue after i made the new exception, what should i do now ?

**Comment by sbromberger** [comment]
> @sbromberger there is a coverage issue after i made the new exception, what should i do now ?

Another reason not to create a new exception - then you have to test it :)

I see you've reverted that change. I still think we need to throw an error if the number of vertices returned exceeds nv(g). That certainly indicates an error and would've caught the original issue.

**Comment by sbromberger** [comment]
I've been testing this against https://www.cs.usfca.edu/~galles/visualization/TopoSortDFS.html this evening and, while we're not getting the *same* ordering, our orderings are valid. Unless @abhinavmehndiratta can come up with a counterexample or another reason we shouldn't merge this, I'd like to do so tomorrow since this is a bugfix and therefore higher-priority.

**Comment by sbromberger** [comment]
closes #1337 
closes #1343 
closes #1341

**Comment by abhinavmehndiratta** [comment]
I think this should work, I've tried this on every testcase I could think of.

**Comment by abhinavmehndiratta** [comment]
I'm still not sure if this is the best way to do this though.

**Comment by sbromberger** [comment]
@abhinavmehndiratta thanks. Let me ask this: would anything change if we passed `visited` to the visitor functins `newvisit`, `visit`, `previsit`, and `postvisit`? We might end up doing this for  BreadthFirst in any case.

**Comment by alimamdouh212** [comment]
> @abhinavmehndiratta thanks. Let me ask this: would anything change if we passed `visited` to the visitor functins `newvisit`, `visit`, `previsit`, and `postvisit`? We might end up doing this for BreadthFirst in any case.
 the problem of the old code that it made multiple dfs calls, every new call to dfs know nothing about the previous calls, so it assumes there is no visited node and reset all visited array to false, if we pass the visited array to visit function, we may make visit equals true if the color is 2, it will fix the problem, but it is just an ugly hack.

**Comment by abhinavmehndiratta** [comment]
Maybe if have a user defined function which takes in the state & returns true/false and only then pushes to the stack. But I don't know if that's a good design.

**Comment by sbromberger** [comment]
OK. We're probably going to be passing `visited` into the `BreadthFirst` visit functions at some point (I think Johnson requires this), so we can revisit (pardon the pun) later. For now, let's get this fix in.

**Comment by sbromberger** [comment]
Just FYI: this code appears to be correct, but it comes at a cost:

v1.3.1
```
julia> @benchmark topological_sort_by_dfs(g)
BenchmarkTools.Trial:
  memory estimate:  69.00 MiB
  allocs estimate:  32
  --------------
  minimum time:     138.424 ms (0.00% GC)
  median time:      142.853 ms (0.00% GC)
  mean time:        143.507 ms (0.93% GC)
  maximum time:     155.655 ms (7.72% GC)
  --------------
  samples:          35
  evals/sample:     1
```

This code:
```
julia> @benchmark LightGraphs.Traversals.topological_sort(g)
BenchmarkTools.Trial:
  memory estimate:  69.50 MiB
  allocs estimate:  39
  --------------
  minimum time:     176.334 ms (0.00% GC)
  median time:      181.164 ms (0.00% GC)
  mean time:        181.673 ms (0.49% GC)
  maximum time:     189.146 ms (3.46% GC)
  --------------
  samples:          28
  evals/sample:     1
```

**Comment by sbromberger** [comment]
Would passing `visited` in allow us to optimize?

**Comment by abhinavmehndiratta** [comment]
I think this is the best we can do. Passing visited would only make it worse.

**Pull request merged**